akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -138,8 +138,16 @@ object GraphStageLogic {
     private[this] var behaviour = initialReceive
 
     /** INTERNAL API */
-    private[akka] def internalReceive(pack: (ActorRef, Any)): Unit =
-      behaviour(pack)
+    private[akka] def internalReceive(pack: (ActorRef, Any)): Unit = {
+      pack._2 match {
+        case Terminated(ref) ⇒
+          if (watching contains ref) {
+            watching -= ref
+            behaviour(pack)
```

Comment:
So you *don't* want to run `behavior(pack)` is the ref is not in the watching set?